src/cmd/services/m3coordinator/downsample/id_pool_types.go

```diff
@@ -125,8 +125,8 @@ func (p *rollupIDProvider) reset(
 	p.index = -1
 	p.newName = newName
 	p.tagPairs = tagPairs
-	p.nameTagIndex = -1
-	p.rollupTagIndex = -1
+	p.nameTagIndex = 0
+	p.rollupTagIndex = 0
```

Comment:
Looks like this makes the conditions `p.nameTagIndex == -1` and `p.rollupTagIndex == -1` in the loop below redundant?

Reply:
hmm actually I think this is now a bug, since the tags will no longer be properly sorted. good catch! I hope a test catches it too!

Reply:
ok that took some serious investigation!

@linasm thanks for catching the bug. unfortunately it didn't break any tests because all it did was change the rollupID (the encoded sorted tags). This would have broken aggregations when deployed, since rollupIDs would have changed for the same metric IDs. Classic backwards compatible issue that is hard to test in CI :(

I added a bunch of tests to to verify how rollupIDs are calculated. The logic is pretty tricky since it tries to merge in our special tags with the metric tags with zero allocations. Unfortunately it was pretty broken  and almost every test case I added was broken. We were just lucky that metric tags typically don't start with `_` so `__name__` and `__rollup__` are usually sorted first.